### DISCUSSION TITLE
Running without Websockts!?

### DISCUSSION THREAD

**Discussion by lnnwvr** [miguelgrinberg/Flask-SocketIO#1855] [Q&A]
I am wondering i have a setup which shouldn work but it does:

I am using waitress as the app server which does not support websockets.

my init looks like this
```
import logging
#import eventlet

from flask import Flask
from flask_socketio import SocketIO
from flask_cors import CORS
from flask_appbuilder import AppBuilder, SQLA

"""
 Logging configuration
"""

logging.basicConfig(format="%(asctime)s:%(levelname)s:%(name)s:%(message)s")
logging.getLogger().setLevel(logging.DEBUG)

app = Flask(__name__)
#eventlet.monkey_patch()
socketio = SocketIO(app, cors_allowed_origins="*", async_mode='threading')
```
When using async mode threading i get this error message but client is receiving events. When i am using asyncmode eventlett it is working too but with a different error message. It works with and without monkey patching.

What is happening here? And what would you recommend? Removing the socketio from init and move it into different script and run just the socket with socketio.run(app=app)?. 

```
The WebSocket transport is not available, you must install a WebSocket server that is compatible with your async mode to enable it. See the documentation for details. (further occurrences of this error will be logged with level INFO)
2022-07-27 12:11:44,950:ERROR:engineio.server:The WebSocket transport is not available, you must install a WebSocket server that is compatible with your async mode to enable it. See the documentation for details. (further occurrences of this error will be logged with level INFO)
```
```
apispec==3.3.2
attrs==21.4.0
Babel==2.10.3
bidict==0.22.0
click==8.1.3
colorama==0.4.5
dnspython==2.2.1
email-validator==1.2.1
eventlet==0.33.1
Flask==2.1.3
Flask-AppBuilder==4.1.3
Flask-Babel==2.0.0
Flask-Cors==3.0.10
Flask-JWT-Extended==4.4.2
Flask-Login==0.6.2
Flask-SocketIO==5.2.0
Flask-SQLAlchemy==2.5.1
Flask-WTF==0.15.1
geojson==2.5.0
greenlet==1.1.2
idna==3.3
itsdangerous==2.1.2
Jinja2==3.1.2
jsonschema==4.7.2
MarkupSafe==2.1.1
marshmallow==3.17.0
marshmallow-enum==1.5.1
marshmallow-sqlalchemy==0.26.1
packaging==21.3
prison==0.2.1
psycopg2-binary==2.9.3
PyJWT==2.4.0
pyparsing==3.0.9
pyrsistent==0.18.1
python-dateutil==2.8.2
python-engineio==4.3.3
python-socketio==5.7.1
pytz==2022.1
PyYAML==6.0
seidr @ git+https://github.com/dttctcs/seidr.git@8b7724ec6278dbf87fdf092b20a54d1a93efb489
six==1.16.0
SQLAlchemy==1.4.39
SQLAlchemy-Utils==0.38.3
waitress==2.1.2
Werkzeug==2.2.0
WTForms==2.3.3
```

**Comment by miguelgrinberg** [comment]
The message is just telling you that WebSocket is not available. If you continue to use waitress there is nothing you can do to address this.

**Reply by lnnwvr** [reply]
Yes and no .. the strange thing it is working. I get events in my react client and i am wondering how.

I was thinking about serving the html stuff with waitress and the websocket on different port with eventlet (running in k8s with ingressroutes and so on)

What do you recommend? Also serving the standard routes with socketio.run? I am not a fan of gunicorn.

**Reply by miguelgrinberg** [reply]
It's not strange that it works. Socket.IO also works over plain HTTP, WebSocket is not required.

I don't know what to recommend. If you prefer to use waitress over having WebSocket, then you don't need my help. If you want to have WebSocket support, then you have to drop Waitress and use one of the supported web servers.

**Comment by lnnwvr** [comment]
Here a error during testing:

```
2022-07-27 14:32:53,442:ERROR:waitress:Exception while serving /socket.io/
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/waitress/channel.py", line 428, in service
    task.service()
  File "/usr/local/lib/python3.10/site-packages/waitress/task.py", line 168, in service
    self.execute()
  File "/usr/local/lib/python3.10/site-packages/waitress/task.py", line 434, in execute
    app_iter = self.channel.server.application(environ, start_response)
  File "/usr/local/lib/python3.10/site-packages/flask/app.py", line 2091, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.10/site-packages/flask_socketio/__init__.py", line 43, in __call__
    return super(_SocketIOMiddleware, self).__call__(environ,
  File "/usr/local/lib/python3.10/site-packages/engineio/middleware.py", line 63, in __call__
    return self.engineio_app.handle_request(environ, start_response)
  File "/usr/local/lib/python3.10/site-packages/socketio/server.py", line 604, in handle_request
    return self.eio.handle_request(environ, start_response)
  File "/usr/local/lib/python3.10/site-packages/engineio/server.py", line 429, in handle_request
    socket = self._get_socket(sid)
  File "/usr/local/lib/python3.10/site-packages/engineio/server.py", line 638, in _get_socket
    raise KeyError('Session is disconnected')
KeyError: 'Session is disconnected'

```

i am totally  confused .. reading the stack trace i would say waitress is somehow serving engineio?

**Reply by lnnwvr** [reply]
Ok .. now i found my answer 

```
The bidirectional channel between the Socket.IO server (Node.js) and the Socket.IO client (browser, Node.js, or [another programming language](https://socket.io/docs/v4/#what-socketio-is)) is established with a [WebSocket connection](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket) whenever possible, and will use HTTP long-polling as
```

so flask socketio supports http long polls ?

Anyway would you recommend splitting waitress flask components and socket components or how would you solve this?

**Reply by miguelgrinberg** [reply]
I don't really know what you are asking me a recommendation for. If you are set on using Waitress, then WebSocket is never going to work. Socket.IO will work, but will use HTTP, which is less performant than WebSocket.

The error is not related to Waitress, I think. The error means that the client is reporting a session id that is unknown or has been disconnected previously.

**Reply by lnnwvr** [reply]
I am thinking about 2 Processes on different porst
serving the html/react stuff with waitress and serving the socket with eventlet on a different process on a different port

Would you recommend this. Or is it safe to serve everything with eventlet. that is what i didn't understand.

**Reply by miguelgrinberg** [reply]
Eventlet is not specific to WebSocket, it has a general purpose web server. If you are okay with using an async framework then it shouldn't be a problem.

On the other side, if you are okay with having two separate processes for the Flask and Socket.IO sides, that's an option too. If you have nginx in front then you can make it all look the same to the client.
